pkg/network/kuryr.go

```diff
@@ -0,0 +1,110 @@
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	operv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-network-operator/pkg/render"
+)
+
+// renderKuryr returns manifests for Kuryr SDN.
+// This creates
+// - the ClusterNetwork object
+// - the kuryr namespace
+// - the kuryr-controller deployment
+// - the kuryr-daemon daemonset
+// and some other small things.
+func renderKuryr(conf *operv1.NetworkSpec, bootstrapData BootstrapData, manifestDir string) ([]*uns.Unstructured, error) {
+	c := conf.DefaultNetwork.KuryrConfig
+	b := bootstrapData.Kuryr
+
+	objs := []*uns.Unstructured{}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+
+	// general kuryr options
+	data.Data["ResourceTags"] = "openshiftClusterID=" + b.ClusterID
+	data.Data["PodSecurityGroups"] = strings.Join(b.PodSecurityGroups, ",")
+	data.Data["WorkerNodesSubnet"] = b.WorkerNodesSubnet
+	data.Data["WorkerNodesRouter"] = b.WorkerNodesRouter
+	data.Data["PodSubnetpool"] = b.PodSubnetpool
+	data.Data["ServiceSubnet"] = b.ServiceSubnet
+	data.Data["OpenStackCloud"] = b.OpenStackCloud
+	// FIXME(dulek): Move that logic to the template once it's known how to dereference pointers there.
+	data.Data["OpenStackInsecureAPI"] = b.OpenStackCloud.Verify != nil && !*b.OpenStackCloud.Verify
+
+	// kuryr-daemon DaemonSet data
+	data.Data["DaemonEnableProbes"] = true
+	data.Data["DaemonProbesPort"] = c.DaemonProbesPort
+
+	// kuryr-controller Deployment data
+	data.Data["ControllerEnableProbes"] = true
+	data.Data["ControllerProbesPort"] = c.ControllerProbesPort
+
+	data.Data["NodeImage"] = os.Getenv("NODE_IMAGE")
+	data.Data["DaemonImage"] = os.Getenv("KURYR_DAEMON_IMAGE")
+	data.Data["ControllerImage"] = os.Getenv("KURYR_CONTROLLER_IMAGE")
+
+	manifests, err := render.RenderDir(filepath.Join(manifestDir, "network/kuryr"), &data)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to render manifests")
+	}
+
+	objs = append(objs, manifests...)
+	return objs, nil
+}
+
+// validateKuryr checks that the Kuryr specific configuration is basically sane.
+func validateKuryr(conf *operv1.NetworkSpec) []error {
+	out := []error{}
+
+	// NOTE(dulek): Dropping this constraint would require changes in Kuryr.
+	if len(conf.ServiceNetwork) != 1 {
+		out = append(out, errors.Errorf("serviceNetwork must have exactly 1 entry"))
+	}
+
+	// TODO(dulek): We should be able to drop this constraint once we test subnetpools with multiple CIDRs.
+	if len(conf.ClusterNetwork) != 1 {
+		out = append(out, errors.Errorf("clusterNetwork must have exactly 1 entry"))
+	}
+
+	return out
+}
+
+// isKuryrChangeSafe currently returns an error if any changes are made.
+// In the future we'll support changing some stuff.
+func isKuryrChangeSafe(prev, next *operv1.NetworkSpec) []error {
+	pn := prev.DefaultNetwork.KuryrConfig
+	nn := next.DefaultNetwork.KuryrConfig
+
+	if reflect.DeepEqual(pn, nn) {
```

Comment:
It would probably be nice to add a TODO to allow whatever safe changes there might be, like probes and the like.